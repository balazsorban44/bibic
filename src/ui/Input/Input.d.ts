import * as React from 'react';
import { StandardProps } from '..';
//TODO:
export interface InputProps extends React.HTMLAttributes<HTMLInputElement> {
  /**The text to show when the input value fails validation */
  error?: string
  label?: string
  name: string
  type?: string,
  placeholder?: string
  required?: boolean
  disabled?: boolean 
  onChange?: Function
  value?: string
  inputProps?: object // REVIEW: rename?
  className?: string
}

declare const Input: React.ComponentType<InputProps>;

export default Input;
