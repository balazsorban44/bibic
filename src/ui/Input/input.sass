@import "../vars"

.input-box
  position: relative
  min-height: 32px
  display: flex
  align-items: baseline
  justify-content: flex-end
  font-family: "Coolvetica", sans-serif
  letter-spacing: .1em
  padding-bottom: 2px
  border-bottom: 2px solid black
  
  label
    padding-right: .5em
    font-weight: bolder
  input
    letter-spacing: .1em
    flex-grow: 1
    padding: .5em
    transition: .3s background-color
    font-family: "Coolvetica", sans-serif
    text-align: right
    font-size: .9em
    border: none
    &::placeholder
      color: transparentize(grey, .6)
  textarea, input
    &:focus
      background-color: transparent !important
      outline: 2px solid $accent-color-dark !important
      color: $dark !important
      &::placeholder
        color: transparent

.input-error
  box-sizing: border-box
  border-bottom: 2px solid transparentize($error, .1)
  animation: shake .5s
  label
    color: transparentize($error, .1)
  // &::after
  //   position: absolute
  //   content: '!'
  //   font-size: .8em
  //   font-weight: bold
  //   font-family: sans-serif
  //   letter-spacing: 0
  //   color: $light
  //   background-color: transparentize($error, .1)
  //   width: 10px
  //   height: 10px
  //   display: flex
  //   justify-content: center
  //   align-items: center
  //   border-radius: 50%
  //   padding: 4px
  //   margin-top: .5em
  //   margin-left: -.5em
  .footnote-asterix::after
    content: ""
  h6
    color: $error
    position: absolute
    top: 8px

.footnote-asterix::after
  content: "*"


@media (min-width: 720px)
  .input-box
    position: relative
    min-width: calc(50% - 2em)
  
@keyframes shake
  0%, 100%
    transform: translateX(0)
  33%, 66%
    transform: translateX(5px)
  50%, 75%
    transform: translateX(-5px)