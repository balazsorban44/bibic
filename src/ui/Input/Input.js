import React, {useState, memo} from "react"
import "./input.sass"

const Input = ({
  component: Component="input",
  error,
  label,
  name,
  type,
  placeholder,
  required,
  disabled,
  onChange,
  value,
  inputProps, // REVIEW: rename?
  className="",
  ...props
}) => {

  const [localValue, setLocalValue] = useState(value)

  const handleChange = ({target: {value}}) => setLocalValue(value)

  const handleBlur = () => onChange({[name]: localValue})

  return (
    <div
      className={`input-box ${className} ${error ? "input-error" : ""}`}
      {...props}
    >
      {label ?
        <label
          className={required ? "footnote-asterix" : ""}
          htmlFor={name}
        >{label}
        </label>
        : null
      }
      <Component
        disabled={disabled}
        name={type || name}
        onBlur={handleBlur}
        onChange={handleChange}
        placeholder={label || placeholder}
        value={localValue}
        {...inputProps}
      />
      {error ? <h6>{error}</h6> : null}
    </div>
  )
}

export default memo(Input)