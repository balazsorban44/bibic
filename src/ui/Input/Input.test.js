import React from 'react'
import {
  render,
  fireEvent,
  cleanup
} from 'react-testing-library'
import 'jest-dom/extend-expect'

import Input from './Input'

afterEach(cleanup)

it("renders correctly", () => {
  expect(render(<Input/>).asFragment()).toMatchSnapshot()
})

it("shows error", () => {
  const {container} = render(<Input error="something"/>)
  expect(container.firstChild.classList).toContain("input-error")
  expect(container.querySelector("h6")).toHaveTextContent("something")
})

it("adds footnote-asterix class when required", () => {
  expect(
    render(
      <Input
        label="something"
        required
      />
    ).container
      .querySelector(".footnote-asterix")
  )
    .toHaveTextContent("something")

  expect(
    render(
      <Input label="something"/>
    )
      .container
      .querySelector("label").classList
  )
    .not.toContain("footnote-asterix")
})

describe("handle events", () => {
  let c
  const changeFn = jest.fn()
  beforeEach(() => {c = render(
    <Input
      name="name"
      onChange={changeFn}
      value="a value"
    />
  )})
  afterEach(cleanup)

  it("handle change", () => {
    fireEvent.change(c.getByValue("a value"), {target: {value: "new value"}})
    expect(c.container.querySelector("input").value).toBe("new value")
  })

  it("blur propagates", () => {
    fireEvent.change(c.getByValue("a value"), {target: {value: "new value"}})
    fireEvent.blur(c.getByValue("new value"))
    expect(changeFn).toBeCalledWith({name: "new value"})
  })
})