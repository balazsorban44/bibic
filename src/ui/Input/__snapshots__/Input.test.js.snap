// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  <div
    class="input-box  "
  >
    <input
      value=""
    />
  </div>
</DocumentFragment>
`;
